Контроль и ограничения (часть 1)

RuslanPro01 opened the pull request:
---
:mortar_board: [Контроль и ограничения (часть 1)](https://up.htmlacademy.ru/react/12/user/854371/tasks/10)

Commit:
feat: Инициализирует store, создаст actions подключит redux к проекту

Commit:
feat: Начнет реализацию табов на главной странице приложения

Commit:
feat: Реализует фильтрацию фильмов по жанрам на главной странице

Commit:
refactor: Внесет правки по замечаниям наставника

Commit:
fix: Удалит импорт неиспользуемого элемента

komaritskaya commented on project/src/common-const.ts:
```diff
@@ -2,7 +2,7 @@ export const Path = {
   Login: '/login',
   FilmsPages: {
-    MainPage: '/films/:id',
+    MainPage: '/AllFilms/:id',
```
В роутах не используется разный регистр (обрати внимание, как это сделано в ТЗ), лучше записать в одно слово или кебаб-кейсом

RuslanPro01 replied:
Это я случайно менял через IDE другую переменную 😅

komaritskaya commented on project/src/components/app/app.tsx:
```diff
@@ -25,11 +27,15 @@ type AppProps = {
-          <Route path={MainPage} element={<Main filmName={filmName} yearFilm={yearFilm} filmGenre={filmGenre} films={films}/>}/>
+          <Route path={MainPage} element={<Main filmName={filmName} yearFilm={yearFilm} filmGenre={filmGenre} films={films}/>}>
+            <Route index element={<FilmCards films={filteredFilms}/>}/>
+            <Route path={':pathGenre'} element={<FilmCards films={filteredFilms}/>}/>
```
Жанр лучше не делать отдельным роутом, просто держать его в стейте приложения и при изменении перерисовывать список фильмов

komaritskaya commented on project/src/components/main-catalog-films/catalog-genres-list.tsx:
```diff
@@ -0,0 +1,36 @@
+  const dispatch = useAppDispatch();
+  const handleGenreLink = (genre: string) => {
+    dispatch(changeGenre(genre));
+    dispatch(setFilmsOfGenre());
```
Эти события лучше разделить - при клике на кнопку жанра, у нас записывается значение активного жанра в стейт. А там, где мы выводим фильмы, мы должны это событие отследить в помощью `useEffect` и вызвать `setFilmsOfGenre`

komaritskaya commented on project/src/components/main-catalog-films/catalog-genres-list.tsx:
```diff
@@ -0,0 +1,36 @@
+import {useAppDispatch, useAppSelector} from '../../hooks';
+import {changeGenre, setFilmsOfGenre} from '../../store/action';
+
+function CatalogGenresList(): JSX.Element {
```
Если это отдельный компонент, то под него должна быть заведена отдельная папка, а то сейчас он лежит в чужой

RuslanPro01 replied:
Этот компонент используется только внутри Компонента `MainPage` поэтому я его не переношу в отдельную папку, он не переиспользуется

komaritskaya commented on project/src/store/reducer.ts:
```diff
@@ -0,0 +1,41 @@
+const initialState: InitialState = {
+  genre: ALL_GENRES,
+  allFilms: films,
+  filteredFilms: films
```
`filteredFilms` лучше не держать в сторе как отдельное значение, т.к. оно формируется из двух других полей. Для таких случаев мы используем селекторы, которые выносим в отдельный файл. Здесь тебе поможет библиотека [reselect](https://github.com/reduxjs/reselect)

Commit:
feat: Подлкючит библиотеку `createSelector` и изменить логику фильтрации фильмов

- Перенесет фильтрацию фильмов в библиотеку с мемоизацией `createSelector`
- Удалить из `reducer.ts` ненужный `action` для фильтрации фильмов из изменения стора.

Commit:
feat: Делает возможным сохранение состояния табов на главной странице main-catalog-films.tsx

Commit:
refactor: Вносит дополнительные правки в проект

Commit:
refactor: Доработает автоскролл

Доработает автоскролл так, чтобы он срабатывал при изменении id в адресной строке. Это позволить добиться автоскролла при переходе в разные карточки фильмов, но не будет скролить при хождении по табам

komaritskaya commented on project/src/utils.ts:
```diff
@@ -0,0 +1,30 @@
+}
+
+function toUpperFirstSubmol(word: string) {
+  word = word.split('').map((symbol, index) => {
```
Тут нехорошая практика: то, что приходит к тебе в аргументах - это внешний скоуп, а переопределяя их, ты его мутируешь. Лучше не переопределять `word`, а использовать промежуточную переменную.

komaritskaya replied:
Сам процесс, вроде бы, называется capitalize (когда мы делаем первую букву заглавной), но это так, тонкости нейминга)
